discoverx/delta_housekeeping.py

```diff
@@ -0,0 +1,368 @@
+                    lambda left, right: left.union(right),
+                    dd_list
+                ),
+                self._spark.sql(statement),
+            )
+
+        errors_df = self._spark.createDataFrame([], self.empty_schema)
+        if errors:
+            errors_df = reduce(
+                lambda left, right: left.union(right),
+                errors
+            )
+
+        out = dh.unionByName(errors_df, allowMissingColumns=True)
+
+        if do_save_as_table:
+            (
+                out
+                .write
+                .format("delta")
+                .mode("append")
+                .option("mergeSchema", "true")
+                .saveAsTable(housekeeping_table_name)
+            )
+
+        return out.toPandas()
+
+
+class DeltaHousekeepingActions:
+    def __init__(
+        self,
+        # delta_housekeeping: DeltaHousekeeping,
+        mapped_pd_dfs: Iterable[pd.DataFrame],
+        # spark: SparkSession = None,
+        min_table_size_optimize: int = 128*1024*1024,  # i.e. 128 MB
+        min_days_not_optimized: int = 7,
+        min_days_not_vacuumed: int = 31,
+        max_optimize_freq: int = 2,
+        max_vacuum_freq: int = 2,
+        small_file_threshold: int = 32*1024*1024,  # i.e. 32 MB
+        min_number_of_files_for_zorder: int = 8,
+        stats: pd.DataFrame = None,  # for testability only
+    ) -> None:
+        # self._delta_housekeeping = delta_housekeeping
+        if stats is None:
+            self._mapped_pd_dfs = mapped_pd_dfs
+            stats = pd.concat(self._mapped_pd_dfs)
+        self._stats: pd.DataFrame = stats
+        # if spark is None:
+        #     spark = SparkSession.builder.getOrCreate()
+        # self._spark = spark
+        self.min_table_size_optimize = min_table_size_optimize
+        self.min_days_not_optimized = min_days_not_optimized
+        self.min_days_not_vacuumed = min_days_not_vacuumed
+        self.max_optimize_freq = max_optimize_freq
+        self.max_vacuum_freq = max_vacuum_freq
+        self.small_file_threshold = small_file_threshold
+        self.min_number_of_files_for_zorder = min_number_of_files_for_zorder
+        self.tables_not_optimized_legend = "Tables that are never OPTIMIZED and would benefit from it"
+        self.tables_not_vacuumed_legend = "Tables that are never VACUUM'ed"
+        self.tables_not_optimized_last_days = "Tables that are not OPTIMIZED often enough"
+        self.tables_not_vacuumed_last_days = "Tables that are not VACUUM'ed often enough"
+        self.tables_optimized_too_freq = "Tables that are OPTIMIZED too often"
+        self.tables_vacuumed_too_freq = "Tables that are VACUUM'ed too often"
+        self.tables_do_not_need_optimize = "Tables that are too small to be OPTIMIZED"
+        self.tables_to_analyze = "Tables that need more analysis (small_files)"
+        self.tables_zorder_not_effective = "Tables for which ZORDER is not being effective"
+
+    def stats(self) -> pd.DataFrame:
```

Comment:
Please remove this method. 
In order to keep the API simple, there should be only one method returning a dataframe, and that method should be `apply` like in `DataExplorerActions`. Please change the implementation of apply to include the diagnostics information as extra boolean columns, see comment below.